### ISSUE TITLE
[Feature Request] InlineCode support for sam local

### ISSUE THREAD

**Issue by j0lly** [aws/aws-sam-cli#790]
Hello there,

I was thinking on taking some work on the topic as InlineCode is a great feature but a few iteration of sam local invoke would be much appreciated during code development.

thus I was planning something along the line of creating a tempfile with InlineCode string content and passing along the context instead the CodeUri path..

Not sure if you see any major issue with the approach (platform, filename (using handler)) otherwise I would happily start implement. 

The hook for the switch case seems reasonably here:
https://github.com/awslabs/aws-sam-cli/blob/1d7df8fce701711963efc32507305c2fdd84a3aa/samcli/commands/local/lib/local_lambda.py#L104

**Comment by sanathkr** [comment]
Hey, thanks for starting this. Definitely a good feature to have. I have commented on the design PR

**Comment by mihaa1** [comment]
Was there any progress with this?

**Comment by rainabba** [comment]
Wow. Could we PLEASE at least get a useful error message for this then? I've wasted most of my day only to come to find the most obvious of all "Hello World" scenarios isn't supported. I just needed an /alive url so I could detect when `sam local start` is responsive. Is there a more appropriate way to do this? Below is my solution for anyone else on a similar journey.

```bash
#!/bin/sh
set -aeux

wait-for-url() {
    echo ":: Testing $1"
    timeout -s TERM 45 bash -c \
    'while [[ "$(curl -s -o /dev/null -L -w ''%{http_code}'' ${0})" != "200" ]];\
    do echo "Waiting for ${0}" && sleep 2;\
    done' ${1}
    echo "OK!"
    curl -I $1
}

trap "exit" INT TERM ERR
trap 'kill $(jobs -p)' EXIT

npm run startApiGateway &

printf "\n:: Waiting for local API Gateway to come up. Initial build/startup takes at least 10 seconds.\n"
sleep 10s
wait-for-url http://127.0.0.1:3000/alive

# Environment ready. Run tests
test/test_local.sh &

wait
```

**Comment by rainabba** [comment]
At [https://github.com/aws/aws-sam-cli/blob/develop/samcli/commands/local/lib/local_lambda.py#L177](https://github.com/aws/aws-sam-cli/blob/develop/samcli/commands/local/lib/local_lambda.py#L177) [`resolve_code_pathresolve_code_path(cwd, codeuri)`](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/utils/codeuri.py) is called, but codeuri will be blank for inline function (see [codeUri](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html#sam-function-codeuri) "_If the PackageType property is set to Zip (default), then one of CodeUri or InlineCode is required_").

As a result (and because no anticipating check first handled things), code breaks at https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/utils/codeuri.py#L43 because the call `os.path.isabs(codeuri)` isn't valid when codeuri is None.

Question now is do we bypass the error (returning None), or should an error be thrown here because inline CANNOT work with local yet (I haven't dug into the mechanics further yet)?

All that's needed to get past the present error:

### fir fix?
```python
    if not codeuri is None :
        if not os.path.isabs(codeuri):
            codeuri = os.path.normpath(os.path.join(cwd, codeuri))
```

### error stack
```bash
Traceback (most recent call last):
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/samcli/local/apigw/local_apigw_service.py", line 317, in _request_handler
    self.lambda_runner.invoke(route.function_name, event, stdout=stdout_stream_writer, stderr=self.stderr)
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/samcli/commands/local/lib/local_lambda.py", line 120, in invoke
    config = self.get_invoke_config(function)
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/samcli/commands/local/lib/local_lambda.py", line 163, in get_invoke_config
    code_abs_path = resolve_code_path(self.cwd, function.codeuri)
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/lib/python3.8/site-packages/samcli/lib/utils/codeuri.py", line 43, in resolve_code_path
    if not os.path.isabs(codeuri):
  File "/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.21.1/libexec/bin/../../../../../opt/python@3.8/lib/python3.8/posixpath.py", line 62, in isabs
    s = os.fspath(s)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

P.S. I've got a fork setup in gitpod and ready for a commit/pr once I'm clear on this first point.

**Comment by specious** [comment]
From a usability perspective, it probably makes sense to detect the condition and say that: "Local invocation of inline code is not yet supported", rather that let it crash with a stack trace.

**Comment by softprops** [comment]
Is the sam team open to prs to add support for this. It seems like it would simple to materialize a file for the online code

**Comment by jfuss** [comment]
We strongly recommend your code to be in a separate file, as this is a general best practice. Other commands will work better without `InlineCode`, such as `sam sync`. We should fail gracefully for `sam local *` commands. We will track that fix in #4505 and will close this as something we are not intending to support.

**Comment by codebased** [comment]
a project full of bugs with no alternative for basic thing to run lambda without frustration and pain.
